design/ceph/object/ceph-bucket-notification-crd.md

```diff
@@ -0,0 +1,91 @@
+
+Extends the rook operator to handle the CRs that would be submitted by users.
+
+## Implementation
+
+The CR for a topic configuration takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephBucketTopic
+metadata:
+  name: # name of the topic
+  namespace: # namespace where topic belongs
+Spec:
+  endpoint: #(mandatory) URI of an endpoint to send push notification to
+  opaqueData: #(optional) opaque data is set in the topic configuration
+  persistent: false #(optional) indication whether notifications to this endpoint are persistent or not (`false` by default)
+  # Endpoint specific parameters
+  http:
+    verifySSL: true #indicate whether the server certificate is validated by the client or not (`true` by default)
+  amqp:
+    ackLevel: broker # none/routable/broker, optional (default - broker)
+    amqpExchange: direct # exchanges must exist and be able to route messages based on topics
+  kafka:
+    useSSL: true # secure connection will be used for connecting with the broker (`false` by default)
+    caLocation: <filepath in rgw pod> # this specified CA will be used, instead of the default one, to authenticate the broker
+    ackLevel: broker # none/broker, optional (default - broker)
+```
+P.S : Endpoint can be of different format depends on the server
+- http -> `http[s]://<fqdn>[:<port]`
+- amqp -> `amqp://[<user>:<password>@]<fqdn>[:<port>][/<vhost>]`
+- kafka -> `kafka://[<user>:<password>@]<fqdn>[:<port]`
+
+The CR for bucket notification takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephBucketNotification
+metadata:
+  name: # name of the notification
+  namespace: # namespace where notification belongs
+Spec:
+  topic: #(mandatory) reference to the topic, topic_arn
+  filter: #(optional) Prefix/Suffix/Regex/Metadata/Tags, optional (default - {})
+    metadata:
+    - name: x-amz-meta-color
+      value: blue
+    stringMatch:
+    - name: prefix
+      value: hello
+    - name: suffix
+      value: .png
+    - name: regex
+      value: [a-z]*
+
+  events: # applicable values [here](https://docs.ceph.com/en/latest/radosgw/s3-notification-compatibility/#event-types), (default all)
+```
+
+After creating resources for notifications, it can be referred in `Additional Config` field of OBC. `Additional Config` is string map of key value pairs and notification can be referred as `bucketNotification` key. It will be list of notification separated by comma.
+
+```yaml
+apiVersion: objectbucket.io/v1alpha1
+kind: ObjectBucketClaim
+metadata:
+  name: ceph-bucket
+spec:
+  bucketName: mybucket
+  storageClassName: rook-ceph-delete-bucket
+  additionalConfig:
+    bucketNotification: "bucketnotification1, bucketnotification2"
+```
+
+```
+P.S Currently OBC's CRDs cannot be modified/updated after creation please [see](https://github.com/kube-object-storage/lib-bucket-provisioner#195), hence bucket notifications could be added only upon the creation of a bucket, and cannot be removed
+```
+
+## Out of Scope
```

Comment:
Even though COSI isn't ready yet, the initial design is completed. I think we need to ensure the bucket notifications also works with COSI or else we might design something that soon needs to be updated anyway for COSI. This makes me think we shouldn't use the additional config field on the OBC and find another way to associate the notification with a bucket. A couple ideas:
1. The BucketClass could have a list of notifications that are to be applied to all the buckets it creates.
1. Create a new CRD that associates the notification with a bucket. 

Option 1 is likely not flexible enough since all buckets aren't likely to require the same notifications. Option 2 is just additional overhead to create the CR instead of just adding a property on the OBC. 

Other ideas?

Reply:
according to the design, the BucketClass has a key value map field called "parameters", which are inherited in the Bucket.
these could be used to set the notifications on the bucket, similar to the current "additionConfig" field in the OBC.
if different sets of bucket require different notifications, they would have to belong to different BucketClasses.

agree that (2) would create unnecessary overhead and dependencies, especially once we support modifications to the configuration.

BTW, the design also scope out things like lifecycle and bucket policies, what would be the approach in these cases?

Reply:
If we use the parameters on the BucketClass, it just limits the admin to configure the notifications, rather than allowing the app to request notifications on the BucketRequest.

Lifecycle and bucket policies seem like they would have the same approach as notifications, we have to decide if the app or the admin needs to control them.

Reply:
On the BucketClass it would make no sense. @travisn is right, it will break the UX (see my other comment), plus in a large environment you may end up with hundreds of BucketClasses to handle all the notification possibilities used by the organization's apps. We surely don't want that.

Reply:
@guimou this makes sense, but I could not see a place where this configuration could be added to the BucketRequest.

Can we use the BucketRequest labels for the notification information?
this would allow us to define a selector in the BucketNotification object, and loosen the coupling between the two objects.

if we go down that path, can we use the same mechanism for OBCs?

Reply:
@travisn : I completely agree with what you say with Bucket Request. But IMO from an application point of OBC will more related to `BucketAccessClass` and `Bucket Access Request`(it is not perfect comparison). For `Bucket Access Request` we can use `parameter` field like OBC at least AFAIR

Reply:
some thoughts about using labels (instead of `additionalConfig`):

1. on the "notification side" we use the same mechanism for OBC and BucketRequest
2. in case of OBC, no need to fix: https://github.com/kube-object-storage/lib-bucket-provisioner/issues/195
3. in case of COSI, allow for good granularity without creating many BucketClasses
4. good manageability:
* when a notification is added, you can use a selector on the label to find all bucket requests that need that notification
* when a BucketRequest/OBC is added you can write a hook that looks into this label and search for this notification

Reply:
@thotz The `parameters` look like they are only on the `BucketAccess`, not the `BucketAccessRequest`. But anyway I don't think associating the access types with the notifications is what we want. The access should really be about managing the access to the buckets, not other behaviors of the bucket. Thus, I would vote to associate the notifications with the BucketRequest. Since there are no `parameters` on the BR, using either labels or annotations should be fine. There is a subtle difference, but since it could be useful to search on the labels, let's go with labels.